include/revng/Pipeline/Location.h

```diff
@@ -0,0 +1,181 @@
+#include "revng/Support/YAMLTraits.h"
+
+namespace pipeline {
+
+namespace concepts {
+
+template<typename Type>
+concept Rank = std::is_same_v<typename Type::RankTag, Type>;
+
+template<typename Type>
+concept RootRank = Rank<Type> && Type::Depth == 0;
+
+namespace detail {
+
+template<typename From, typename To>
+struct RechabilityHelper {
+private:
+  static constexpr bool Found = std::is_same_v<From, To>;
+  static_assert(Rank<From> && Rank<To>);
+  using Next = std::conditional_t<(RootRank<From> || From::Depth <= To::Depth),
+                                  std::false_type,
+                                  RechabilityHelper<typename From::Parent, To>>;
+
+public:
+  static constexpr bool value = Found || Next::value;
+};
+
+} // namespace detail
+
+// clang-format off
+template<typename First, typename Second>
+concept ConvertibleRank = Rank<First>
+                          && Rank<Second>
+                          && detail::RechabilityHelper<Second, First>::value;
+// clang-format on
+
+} // namespace concepts
+
+template<concepts::Rank RankDefinition>
+class Location : public RankDefinition::Tuple {
+  friend RankDefinition;
+
+private:
+  using Tuple = typename RankDefinition::Tuple;
+  static_assert(std::tuple_size_v<Tuple> == RankDefinition::Depth);
+  static constexpr std::size_t Size = RankDefinition::Depth;
+
+private:
+  constexpr static std::string_view Separator = "/";
+
+public:
+  using Tuple::Tuple;
+  Location(const Location &) = default;
+  Location(Location &&) = default;
+  Location &operator=(const Location &) = default;
+  Location &operator=(Location &&) = default;
+
+  Tuple &tuple() { return *this; }
+  const Tuple &tuple() const { return *this; }
+
+  template<concepts::ConvertibleRank<RankDefinition> AnotherRank>
+  auto &get(const AnotherRank &Rank) {
+    return std::get<AnotherRank::Depth - 1>(tuple());
+  }
+  template<concepts::ConvertibleRank<RankDefinition> AnotherRank>
+  const auto &get(const AnotherRank &Rank) const {
+    return std::get<AnotherRank::Depth - 1>(tuple());
+  }
+
+  // clang-format off
+  template<typename AnotherRank>
+    requires (concepts::ConvertibleRank<RankDefinition, AnotherRank>
+              || concepts::ConvertibleRank<AnotherRank, RankDefinition>)
+  static constexpr Location<RankDefinition>
+  convert(Location<AnotherRank> &&Another) {
+    // clang-format on
+    Location<RankDefinition> Result;
+
+    constexpr auto Common = std::min(RankDefinition::Depth, AnotherRank::Depth);
+    constexprRepeat<Common>([&Result, &Another]<std::size_t I> {
+      std::get<I>(Result.tuple()) = std::move(std::get<I>(Another.tuple()));
+    });
+
+    return Result;
+  }
+
+  // clang-format off
+  template<typename AnotherRank>
+    requires (concepts::ConvertibleRank<RankDefinition, AnotherRank>
+              || concepts::ConvertibleRank<AnotherRank, RankDefinition>)
+  static constexpr Location<RankDefinition>
+  convert(const Location<AnotherRank> &Another) {
```

Comment:
Why do we need the `Location<AnotherRank> &&Another` version?

Reply:
What if there are ever locations with expensive to copy keys? Or, maybe, keys that cannot be copied at all?

Reply:
In most cases they will be intergers or `MetaAddress` (designed to be cheaply copyable).
I see more liability in duplicating the code (you end  updating only one of the two implementations) than an actual benefit in terms of efficiency.

Reply:
Once we move to c++23 these can be merged together using the new `this type deduction` trickery. Until then, I'm fine with either keeping a move-only copy or removing it outright.

Reply:
Drop then. Can you point me to some this type deduction resource?
Does clang 14 supports it? `:P`

Reply:
As of now, only msvc 19.32 (vs 17.2) and EDG's compiler support that, sadly. But it's to be expected, I guess, the c++23 is not even finalized yet.

Here's the relevant proposal paper: https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p0847r7.html

Reply:
Gone.